### PR TITLE
task/move-hydration-to-serializer

### LINKED ISSUE #17: Create IdentityMapping

### LINKED ISSUE #18: Move Hydration to Serializer

### PR DESCRIPTION
**Pull request by DanielBadura**
…working atm)

fixes #18 fixes #17

### CODE REVIEW

**Comment by MarcoGiesen** [comment]
Well your code is generating some exceptions, because of the DateTime-Format while executing unserialization. You could add this:

```
        $this->serializer = SerializerBuilder::create()
            ->configureHandlers(function (HandlerRegistry $registry) {
                $registry->registerSubscribingHandler(new IssueHandler($this));
            })
            ->addDefaultDeserializationVisitors()
            ->addDefaultHandlers()
            ->addDefaultListeners()
            ->build();
```

in the client constructor to add some default handlers and stuff so it works smoothly.

**Comment by DanielBadura** [comment]
@MarcoGiesen @DavidBadura now everything should be fine and working.
Do you guys have some objections or sth?

**Comment by MarcoGiesen** [comment]
I think something went wrong while moving the id mapper in the repositories. At commit 494eba8 everything is working fine, but after d86f5ea the hydrate-function is called in an infinity loop.

File: src/Hydration/Hydration.php [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        }
+
+        $factory           = new LazyLoadingValueHolderFactory();
+
+        $initializer = function (
+            &$wrappedObject,
+            LazyLoadingInterface $proxy,
+            $method
+        ) use ($user, $userRepository) {
+            if ($method == 'getId' || $method == 'getName') {
```
Comment (MarcoGiesen) [inline src/Hydration/Hydration.php:92]: remove getName, if you load an issue only the user-id is returned nothing else

File: src/Hydration/Hydration.php [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    {
+        $userRepository = $this->client->getUserRepository();
+
+        $userStruct = $this->getIdentity($user, $userRepository);
+
+        if ($userStruct !== null) {
+            return $userStruct;
+        }
+
+        $factory           = new LazyLoadingValueHolderFactory();
```
Comment (MarcoGiesen) [inline src/Hydration/Hydration.php:85]: code format?

File: src/Hydration/Hydration.php [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    {
+        $projectRepository = $this->client->getProjectRepository();
+
+        $projectStruct = $this->getIdentity($project, $projectRepository);
+
+        if ($projectStruct !== null) {
+            return $projectStruct;
+        }
+
+        $factory           = new LazyLoadingValueHolderFactory();
```
Comment (MarcoGiesen) [inline src/Hydration/Hydration.php:45]: code format?

File: src/Proxy/User.php [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+     */
+    public function getId()
+    {
+        return $this->id;
+    }
+
+    /**
+     * @return string
+     */
+    public function getName()
```
Comment (MarcoGiesen) [inline src/Proxy/User.php:38]: remove getName and the property, because it does not exist in redmine at the user endpoint

Comment (DanielBadura) [inline src/Proxy/User.php:38]: it does when you get an issue. 
look here: https://github.com/DanielBadura/Redmine-Api-Client/pull/19/files#diff-07e777d2c4e7220582f6d5a5d4c3483bR8

Comment (MarcoGiesen) [inline src/Proxy/User.php:38]: i justed checked and dumped an issue in the current master, there is no name property.. just id, login, firstname, lastname, mail, auth_source_id, memberships and groups

Comment (MarcoGiesen) [inline src/Proxy/User.php:38]: i just checked with the api, there actually should be a name attribute. But our user struct does not take this value at the moment, so when i dumped an issue there is no name.

File: src/Hydration/Hydration.php (hunk 3 above)
Comment (DanielBadura) [inline src/Hydration/Hydration.php:45]: i'll do this in the next pr. Then i'll format everything

Comment (MarcoGiesen) [inline src/Hydration/Hydration.php:45]: sounds good.

File: src/Proxy/User.php (hunk 4 above)
Comment (DanielBadura) [inline src/Proxy/User.php:38]: i added now a `getName()` method to our user struct too

Comment (MarcoGiesen) [inline src/Proxy/User.php:38]: ok, now it seems to be working fine

**Comment by tolry** [comment]
A few questions / notes concerning your implementation :smile: 
- I would turn the naming around
  - hydration = this is an active component, it takes data (arrays) and transforms it into an object, therefore I might have called it hydrator
  - mapper = this is not an active component, but rather a simple data structure holding objects, therefore I would have called it simply IdentityMap (http://martinfowler.com/eaaCatalog/identityMap.html)
- what are the advantages of having multiple maps over just one, holding all objects?
- what are the advantages of having a class for the map instead of a native array/map which would be encapsulated in the hydrator or, e.g. an entity manager?
- there is no way to clear the identity map yet, in order to refresh an object at runtime, right?

**Comment by DanielBadura** [comment]
Yeah, you're right about the Naming of `Hydration` => `Hydrator` and `IndentityMapper` => `IdentityMap`.
The advantage is in my opinion that I dont have to prefix the ID that I use to save the Entity. If i would only have one map, then i must prefix it with sth like `$map->store('project.' . $project->getId());`.
And to the point extra class: i thought it would be better to have this seperated.
To the last point, yes you are right. I wanted to add this later on: #21

**Comment by tolry** [comment]
About a single identity map vs. multiple: you could simply prefix the ID with `get_class()`, which would be a generic solution, without additional work per entity class. Having only one identity map also has the advantage of being able to clear it in one swoop - not having to iterate over all maps / repositories to clear everything.

**Pull request merged**